### PR TITLE
Refactor c8y-config-plugin to use c8y-config-manager actor

### PR DESCRIPTION
**Pull request by albinsuresh**
## Proposed changes

- [x] Refactor c8y-config-plugin to use c8y-config-manager actor
- [x] Port existing tests into actor based tests
- [ ] Update existing rust integration tests to use all necessary library actors

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: plugins/c8y_configuration_plugin/src/main.rs [hunk 1]
```diff
@@ -73,124 +63,99 @@ pub struct ConfigPluginOpt {
...
     let tedge_config_location =
         tedge_config::TEdgeConfigLocation::from_custom_root(&config_plugin_opt.config_dir);
-    let config_repository = tedge_config::TEdgeConfigRepository::new(tedge_config_location.clone());
-
     let log_level = if config_plugin_opt.debug {
         tracing::Level::TRACE
     } else {
         get_log_level(
-            "c8y-configuration-plugin",
+            "c8y-log-plugin",
```
Comment (didier-wenzek) [inline plugins/c8y_configuration_plugin/src/main.rs:77]: Strange as this is an update of the *configuration* plugin.

File: plugins/c8y_configuration_plugin/src/main.rs [hunk 2]
```diff
@@ -73,124 +63,99 @@ pub struct ConfigPluginOpt {
...
+        C8YHttpProxyBuilder::new(c8y_http_config, &mut http_actor, &mut jwt_actor);
+
+    let mut fs_watch_actor = FsWatchActorBuilder::new();
+    let mut signal_actor = SignalActor::builder();
+    let mut timer_actor = TimerActor::builder();
+
+    //Instantiate health monitor actor
+    let mut health_actor = HealthMonitorBuilder::new(PLUGIN_NAME);
+    let mqtt_config = health_actor.set_init_and_last_will(mqtt_config);
+    let mut mqtt_actor = MqttActorBuilder::new(mqtt_config.clone().with_session_name(PLUGIN_NAME));
```
Comment (didier-wenzek) [inline plugins/c8y_configuration_plugin/src/main.rs:113]: As this has to done for any main function using the health actor, one can consider in a follow-up task to provide an `HealthMonitorBuilder` method that connect to the MQTT actor updating the MQTT config.

File: crates/core/tedge_actors/src/test_helpers.rs [hunk 3]
```diff
@@ -309,7 +312,7 @@ where
 /// returning None, when no message is received after a given duration.
 pub struct TimedBoxReceiver<T> {
     timeout: Duration,
-    inner: T,
+    pub inner: T,
```
Comment (didier-wenzek) [inline crates/core/tedge_actors/src/test_helpers.rs:315]: Why is this required?

**Review by didier-wenzek — COMMENTED** [review]
I would be happy to approved. I just wonder why the system tests are failing.

File: crates/extensions/c8y_http_proxy/src/actor.rs [hunk 4]
```diff
@@ -308,12 +306,24 @@ impl C8YHttpProxyActor {
             download_info.auth = Some(Auth::new_bearer(token.as_str()));
         }
 
-        // Download a file to tmp dir
-        let file_name = request.file_path.file_name().unwrap().to_str().unwrap();
-        let downloader: Downloader = Downloader::new_sm(file_name, &None, self.tmp_dir.clone());
+        // Download file to the target directory with a temp name
+        let file_name = nanoid!();
```
Comment (albinsuresh) [inline crates/extensions/c8y_http_proxy/src/actor.rs:310]: @reubenmiller suggested that this name be something deterministic. I'm thinking of using the target file name itself suffixed with a `.tmp` so that it's easier to identify even if those stay around due to some crashes/bugs.

File: plugins/c8y_configuration_plugin/src/main.rs [hunk 5]
```diff
@@ -1,46 +1,36 @@
...
-#[cfg(test)]
-mod tests;
-
-use crate::config::PluginConfig;
-
-use anyhow::Result;
-use c8y_api::http_proxy::C8YHttpProxy;
-use c8y_api::http_proxy::JwtAuthHttpProxy;
+// #[cfg(test)]
+// mod tests;
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply plugins/c8y_configuration_plugin/src/main.rs:2]: All those tests using test mqtt broker and mock http server were ported into the actor module as pure actor based unit tests with simple message boxes. I still wanted to keep just one of these tests to really make sure that the real interaction between mqtt and http actors are also tested with the help of the same test mqtt broker and mock http server. That's why I didn't remove it initially. But later it slipped my mind. I'll take another attempt to preserve one test. Will remove this module completely if  I don't succeed.

File: crates/core/tedge_actors/src/test_helpers.rs (hunk 3 above)
Comment (didier-wenzek) [reply crates/core/tedge_actors/src/test_helpers.rs:315]: > Needed this to access `send` messages using the message box.

Okay. I added `impl AsMut<T> for TimedBoxReceiver<T>` for that purpose. But `pub` works too.

> I can implement a `send()` method for this `TimedBoxReceiver` as well and delegate that to this `inner`.

Yes, that would be better than both `pub` and `AsRef`. You will need to add a constraint that `T` is a `Sender` ... and to `impl Sender for SimpleMessageBox`.

File: tests/RobotFramework/tests/config_management/child_conf_mgmt_plugin.robot [hunk 6]
```diff
@@ -181,7 +181,7 @@ No response from child device on snapshot request
     Should Be Equal    @{listen}    {"url":"http://${PARENT_IP}:${HTTP_PORT}/tedge/file-transfer/${CHILD_SN}/config_snapshot/config1","path":"/home/pi/config1","type":"config1"}
 
     #CHECK TIMEOUT MESSAGE
-    Cumulocity.Operation Should Be FAILED    ${operation}    failure_reason=Timeout due to lack of response from child device: ${CHILD_SN} for config type: config1    timeout=30
+    Cumulocity.Operation Should Be FAILED    ${operation}    failure_reason=Timeout due to lack of response from child device: ${CHILD_SN} for config type: config1    timeout=60
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply tests/RobotFramework/tests/config_management/child_conf_mgmt_plugin.robot:184]: The operation timeout was increased to 60s.

File: crates/extensions/c8y_http_proxy/src/actor.rs [hunk 7]
```diff
@@ -308,12 +306,24 @@ impl C8YHttpProxyActor {
...
+                "Parent directory of {:?} not found",
+                request.file_path
+            ))
+        })?;
+        debug!(target: self.name(), "Downloading from: {:?}", download_info.url());
+        let downloader: Downloader = Downloader::new_sm(&file_name, &None, parent_dir);
         downloader.download(&download_info).await?;
 
         // Move the downloaded file to the final destination
+        debug!(
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/extensions/c8y_http_proxy/src/actor.rs:322]: It'd be helpful to know the context. The `move_file` util API could be called from different actors. So, if we just rely on the log from that level, we wouldn't know who performed the move. So, logging from here adds that context that the downloader is the one performing this specific file move.

File: crates/extensions/c8y_http_proxy/src/actor.rs [hunk 8]
```diff
@@ -308,12 +306,24 @@ impl C8YHttpProxyActor {
...
+            "Moving downloaded file from {:?} to {:?}",
+            downloader.filename(),
+            request.file_path
+        );
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/extensions/c8y_http_proxy/src/actor.rs:308]: > However, after downloading a file, 99% of usecases requires to move a file from location A to B.

I would even say 100% of the time as otherwise one has no clue is a file is the result of a complete download or some garbage left by an aborted download.

**Review by didier-wenzek — COMMENTED** [review]
Thank you! I'm surprised by the amount of work - notably to port the tests. Estimating the task, I was expecting something more straightforward. I was wrong. We have to keep this in mind for the refactoring of the c8y mapper. I'm afraid the complexity will be similar.

I will approve once the system infra back.
* Take a decision for plugins/c8y_configuration_plugin/src/test.rs
* Fix the recurrent typo in crates/extensions/c8y_config_manager/src/tests.rs

File: crates/extensions/c8y_http_proxy/src/actor.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/extensions/c8y_http_proxy/src/actor.rs:313]: Deferred to a follow-up PR to avoid touching other components using the downloader (tedge-agent and firmware-plugin).

**Review by didier-wenzek — APPROVED** [review]
Approved
